src/main/scala-3/com/raquo/airstream/split/SplittableTypeMacros.scala

```diff
@@ -0,0 +1,338 @@
+package com.raquo.airstream.split
+
+import com.raquo.airstream.core.{
+  EventStream,
+  Signal,
+  Observable,
+  BaseObservable
+}
+import scala.quoted.{Expr, Quotes, Type}
+
+/** `SplittableTypeMacros` turns this code
+  *
+  * ```scala
+  * sealed trait Foo
+  * final case class Bar(strOpt: Option[String]) extends Foo
+  * enum Baz extends Foo {
+  *   case Baz1, Baz2
+  * }
+  * case object Tar extends Foo
+  * val splitter = fooSignal.splitMatch
+  *   .handleCase { case Bar(Some(str)) => str } { (str, strSignal) =>
+  *     renderStrNode(str, strSignal)
+  *   }
+  *   .handleCase { case baz: Baz => baz } { (baz, bazSignal) =>
+  *     renderBazNode(baz, bazSignal)
+  *   }
+  *   .handleCase {
+  *     case Tar    => ()
+  *     case _: Int => ()
+  *   } { (_, _) => div("Taz") }
+  *   .toSignal
+  * ```
+  *
+  * into this code:
+  *
+  * ```scala
+  * val splitter = fooSignal.
+  *  .map { i =>
+  *    i match {
+  *      case Bar(Some(str)) => (0, str)
+  *      case baz: Baz => (1, baz)
+  *      case Tar => (2, ())
+  *      case _: Int => (2, ())
+  *    }
+  *  }
+  *  .splitOne(_._1) { ??? }
+  * ```
+  *
+  * After macros expansion, compiler will warns above code "match may not be
+  * exhaustive" and "unreachable case" as expected.
+  */
+object SplittableTypeMacros {
+
+  extension [Self[+_] <: Observable[_], I](inline observable: BaseObservable[Self, I]) {
+    inline def splitMatch: MatchSplitObservable[Self, I, Nothing] =
+      MatchSplitObservable.build(
+        observable,
+        Nil,
+        Map.empty[Int, Function[Any, Nothing]]
+      )
+  }
+
+  extension [Self[+_] <: Observable[_], I, O](
+    inline matchSplitObservable: MatchSplitObservable[Self, I, O]
+  ) {
+    inline def handleCase[A, B, O1 >: O](inline casePf: PartialFunction[A, B])(inline handleFn: ((B, Signal[B])) => O1) = ${
+      handleCaseImpl('{ matchSplitObservable }, '{ casePf }, '{ handleFn })
+    }
+
+    inline private def handlePfType[T](inline casePf: PartialFunction[Any, T]) = ${
+      handleTypeImpl[Self, I, O, T]('{ matchSplitObservable }, '{ casePf })
+    }
+
+    inline def handleType[T]: MatchTypeObservable[Self, I, O, T] = handlePfType[T] { case t: T => t }
```

Comment:
The reason I need to take this "detour" is that, originally this should be `handleType[T][O1 >:O](inline handleFn: ((T, Signal[T])) => O1)`. `O1>:O` here is very important because it helps refining the output type from the ground "Nothing" up to the output type we want. Sadly Scala 3 right now can only do `handleType[T, O1 >:O]`, and this force user to explicitly named the output type of the `handleFn` function.
Until [SIP-47](https://docs.scala-lang.org/sips/clause-interleaving.html) is widely adopted, we will have to stick into this detour

Reply:
Do I understand correctly that calling `handleType[Foo[A]]` is equivalent to the user writing `handleType { case f: Foo[A @unchecked] => f }` – that is, the `Foo` will be checked, but `A` will be unchecked? That's fine and that's what I would expect, given type erasure, just wanted to double check.

Reply:
Yeah I believe so, the end result of this macro will create a simple `match/cases` code, before any transformation happens. So it should follow the normal behavior of scala